main/src/ca/uwaterloo/flix/language/phase/Deriver.scala

```diff
@@ -167,6 +170,121 @@ object Deriver extends Phase[KindedAst.Root, KindedAst.Root] {
       KindedAst.MatchRule(pat, guard, exp)
   }
 
+  /**
+    * Creates a toString instance for the given enum.
+    *
+    * {{{
+    * enum E[a] with Eq {
+    *   case C1
+    *   case C2(a)
+    *   case C3(a, a)
+    * }
+    * }}}
+    *
+    * yields
+    *
+    * {{{
+    * instance Eq[E[a]] with Eq[a] {
+    *   pub def eq(x: E[a], y: E[a]): Bool = (x, y) match {
+    *     case (C0, C0) => true
+    *     case (C1(x0), C1(y0)) => x0 == y0
+    *     case (C2(x0, x1), C2(y0, y1)) => x0 == y0 and x1 == y1
+    *     case _ => false
+    *   }
+    * }
+    * }}}
+    */
+  private def createEq(enum: KindedAst.Enum, loc: SourceLocation, root: KindedAst.Root)(implicit flix: Flix): KindedAst.Instance = enum match {
+    case KindedAst.Enum(_, _, _, tparams, _, cases, _, sc, _) =>
+      val eqClassSym = PredefinedClasses.lookupClassSym("Eq", root)
+      val eqDefSym = Symbol.mkDefnSym("Eq.eq")
+
+      // create a match rule for each case and put them in a match expression
+      val mainMatchRules = cases.values.map(createEqMatchRule(_, loc, root))
+      val defaultRule = KindedAst.MatchRule(KindedAst.Pattern.Wild(Type.freshVar(Kind.Star, loc), loc), KindedAst.Expression.True(loc), KindedAst.Expression.False(loc))
+      val varSym1 = Symbol.freshVarSym("x", loc)
+      val varSym2 = Symbol.freshVarSym("y", loc)
+      val exp = KindedAst.Expression.Match(
+        KindedAst.Expression.Tuple(List(mkVarExpr(varSym1, loc), mkVarExpr(varSym2, loc)), loc),
+        (mainMatchRules ++ List(defaultRule)).toList,
+        loc
+      )
+
+      val spec = KindedAst.Spec(
+        doc = Ast.Doc(Nil, loc),
+        ann = Nil,
+        mod = Ast.Modifiers.Empty,
+        tparams = tparams,
+        fparams = List(KindedAst.FormalParam(varSym1, Ast.Modifiers.Empty, sc.base, loc), KindedAst.FormalParam(varSym2, Ast.Modifiers.Empty, sc.base, loc)),
+        sc = Scheme(
+          tparams.map(_.tpe),
+          List(Ast.TypeConstraint(eqClassSym, sc.base, loc)),
+          Type.mkPureUncurriedArrow(List(sc.base, sc.base), Type.mkBool(loc), loc)
+        ),
+        tpe = Type.mkBool(loc),
+        eff = Type.Cst(TypeConstructor.True, loc),
+        loc = loc
+      )
+      val defn = KindedAst.Def(eqDefSym, spec, exp)
+
+      // Add a type constraint to the instance for any non-wild param with kind Star
+      val tconstrs = tparams.collect {
+        case tparam if tparam.tpe.kind <:: Kind.Star && !tparam.name.isWild => Ast.TypeConstraint(eqClassSym, tparam.tpe, loc)
+      }
+
+      KindedAst.Instance(
+        doc = Ast.Doc(Nil, loc),
+        mod = Ast.Modifiers.Empty,
+        sym = eqClassSym,
+        tpe = sc.base,
+        tconstrs = tconstrs,
+        defs = List(defn),
+        ns = Name.RootNS,
+        loc = loc
+      )
+  }
+
```

Comment:
I find the code a bit hard to follow. It would be nice if the comments had some more of the fragments.

Reply:
I added some more documentation especially to the match/case part. Let me know if more is needed.